### ISSUE TITLE
wrap_kvs will wrap the instance but self of instance is not wrapped

### ISSUE THREAD

**Issue by andeaseme** [i2mint/dol#18]
This behavior is discovered while trying to display available options in a dropdown with streamlitfront without accessing the actual value until it is needed.

The get item of an instance is different from the get item from the `self` within the instance's method when wrapping a class with `wrap_kvs`.  The `self` behaves as if the instance is not wrapped.

The test below has two classes: 
1. DoubleWrapSquareStore reapplies `wrap_kvs` to `self` in the `item_getter` method call.
1. SingleWrapSquareStore does not reapply `wrap_kvs` and acts like it is not wrapped but you would expect the instance and self to have the same behavior.

Test:
```
import math
from dataclasses import dataclass
from typing import Any

from dol import wrap_kvs


@dataclass
class ItemGetter:
    """Wraps value in a callable to be retrieved later"""

    instance: Any
    key: Any

    def __call__(self):
        return self.instance[self.key]


wrap_kvs_as_square_data = wrap_kvs(data_of_obj=lambda x: x * x, obj_of_data=math.sqrt)


@wrap_kvs_as_square_data
class DoubleWrapSquareStore(dict):
    def item_getter(self, key) -> ItemGetter:
        """The instance self has wrap_kvs applied again to retain the data conversion"""
        return ItemGetter(instance=wrap_kvs_as_square_data(self), key=key)


@wrap_kvs_as_square_data
class SingleWrapSquareStore(dict):
    def item_getter(self, key) -> ItemGetter:
        """The instance self does not have wrap_kvs reapplied and does not act like the instance"""
        return ItemGetter(instance=self, key=key)


def test_square_store(ss):
    ss['2'] = 2
    ig_from_instance = ItemGetter(ss, '2')()
    ig_from_self = ss.item_getter('2')()

    print(
        f'\nTesting {type(ss).__name__}'
        'These values should be equal:',
        f"{ss['2']=}",
        f'{ig_from_instance=}',
        f'{ig_from_self=}',
        sep='\n',
    )

    assert (
        ss['2'] == ig_from_instance
    ), f'{type(ss).__name__}: ItemGetter value from instance does not match actual value'

    ig_2 = ss.item_getter('2')
    assert (
        ss['2'] == ig_from_self
    ), f'{type(ss).__name__}: ItemGetter value from self does not match actual value'

    assert (
        ig_from_instance == ig_from_self
    ), f'{type(ss).__name__}: ItemGetter value from instance does not match ItemGetter value from self'


test_square_store(DoubleWrapSquareStore())
test_square_store(SingleWrapSquareStore())

```
Result:
```
============================= test session starts ==============================
collecting ... 
Testing DoubleWrapSquareStoreThese values should be equal:
ss['2']=2.0
ig_from_instance=2.0
ig_from_self=2.0

Testing SingleWrapSquareStoreThese values should be equal:
ss['2']=2.0
ig_from_instance=2.0
ig_from_self=4

plunk/ap/pain/wrap_kvs_and_self.py:None (plunk/ap/pain/wrap_kvs_and_self.py)
wrap_kvs_and_self.py:64: in <module>
    test_square_store(SingleWrapSquareStore())
wrap_kvs_and_self.py:54: in test_square_store
    assert (
E   AssertionError: SingleWrapSquareStore: ItemGetter value from self does not match actual value
E   assert 2.0 == 4






collected 0 items / 1 error

=============================== 1 error in 0.13s ===============================
```

**Comment by thorwhalen** [comment]
Slightly editted:

```python
import math
from dataclasses import dataclass
from typing import Any

from dol import wrap_kvs


@dataclass
class ItemGetter:
    """Wraps value in a callable to be retrieved later"""

    instance: Any
    key: Any

    def __call__(self):
        return self.instance[self.key]


wrap_kvs_as_square_data = wrap_kvs(data_of_obj=lambda x: x * x, obj_of_data=lambda x: math.sqrt(x))


@wrap_kvs_as_square_data
class DoubleWrapSquareStore(dict):
    def item_getter(self, key) -> ItemGetter:
        """The instance self has wrap_kvs applied again to retain the data conversion"""
        return ItemGetter(instance=wrap_kvs_as_square_data(self), key=key)


@wrap_kvs_as_square_data
class SingleWrapSquareStore(dict):
    def item_getter(self, key) -> ItemGetter:
        """The instance self does not have wrap_kvs reapplied and does not act like the instance"""
        return ItemGetter(instance=self, key=key)


def test_square_store(ss):
    ss['2'] = 2
    ig_from_instance = ItemGetter(ss, '2')()
    ig_from_self = ss.item_getter('2')()

    print(
        f'\nTesting {type(ss).__name__}'
        'These values should be equal:',
        f"{ss['2']=}",
        f'{ig_from_instance=}',
        f'{ig_from_self=}',
        sep='\n',
    )

    assert (
        ss['2'] == ig_from_instance
    ), f'{type(ss).__name__}: ItemGetter value {ig_from_instance=} from instance does not match actual value {ss["2"]=}'

    ig_2 = ss.item_getter('2')
    assert (
        ss['2'] == ig_from_self
    ), f'{type(ss).__name__}: ItemGetter value {ig_from_self=} from self does not match actual value {ss["2"]=}'

    assert (
        ig_from_instance == ig_from_self
    ), f'{type(ss).__name__}: ItemGetter value {ig_from_instance=} from instance does not match ItemGetter value from self {ig_from_self=}'


test_square_store(DoubleWrapSquareStore())
test_square_store(SingleWrapSquareStore())
```

```
Testing DoubleWrapSquareStoreThese values should be equal:
ss['2']=2.0
ig_from_instance=2.0
ig_from_self=2.0

Testing SingleWrapSquareStoreThese values should be equal:
ss['2']=2.0
ig_from_instance=2.0
ig_from_self=4
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
[/Users/thorwhalen/Dropbox/py/notebooks/tw/003](https://file+.vscode-resource.vscode-cdn.net/Users/thorwhalen/Dropbox/py/notebooks/tw/003) - Scrap 2023.ipynb Cell 9 line 6
     [59](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=58)     assert (
     [60](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=59)         ig_from_instance == ig_from_self
     [61](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=60)     ), f'{type(ss).__name__}: ItemGetter value {ig_from_instance=} from instance does not match ItemGetter value from self {ig_from_self=}'
     [64](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=63) test_square_store(DoubleWrapSquareStore())
---> [65](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=64) test_square_store(SingleWrapSquareStore())

[/Users/thorwhalen/Dropbox/py/notebooks/tw/003](https://file+.vscode-resource.vscode-cdn.net/Users/thorwhalen/Dropbox/py/notebooks/tw/003) - Scrap 2023.ipynb Cell 9 line 5
     [50](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=49) assert (
     [51](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=50)     ss['2'] == ig_from_instance
     [52](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=51) ), f'{type(ss).__name__}: ItemGetter value {ig_from_instance=} from instance does not match actual value {ss["2"]=}'
     [54](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=53) ig_2 = ss.item_getter('2')
     [55](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=54) assert (
---> [56](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=55)     ss['2'] == ig_from_self
     [57](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=56) ), f'{type(ss).__name__}: ItemGetter value {ig_from_self=} from self does not match actual value {ss["2"]=}'
     [59](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=58) assert (
     [60](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=59)     ig_from_instance == ig_from_self
     [61](vscode-notebook-cell:/Users/thorwhalen/Dropbox/py/notebooks/tw/003%20-%20Scrap%202023.ipynb#Z2344sZmlsZQ%3D%3D?line=60) ), f'{type(ss).__name__}: ItemGetter value {ig_from_instance=} from instance does not match ItemGetter value from self {ig_from_self=}'

AssertionError: SingleWrapSquareStore: ItemGetter value ig_from_self=4 from self does not match actual value ss["2"]=2.0
```

**Comment by thorwhalen** [comment]
**Verified reproducible on HEAD** (2026-07). Same root cause as #9: a `wrap_kvs`-decorated class that calls `self[k]` internally hits the **unwrapped** inner instance, so the transform pipeline is bypassed for those in-class accesses (the `SingleWrapSquareStore` case in the issue returns `4` instead of `2.0`).

**Triage:** `priority: high`, Wave 1 — resolve together with #9/#12. Either the `FirstArgIsMapping` marker fix (#12) covers it, or we standardize + document the "re-wrap `self`" pattern (the `DoubleWrapSquareStore` workaround). This is already listed as a known limitation in the repo's `CLAUDE.md`. Mechanics: `misc/docs/dol_architecture_map.md` §5.4.

**Comment by thorwhalen** [comment]
**Verified distinct from #9** — the #9/#12 fix (#72) does *not* resolve this. Root cause confirmed in code: `Store.wrap` wraps by **delegation (has-a)**, not inheritance. `delegate_to` (`dol/base.py`) installs the wrapped class's methods as `DelegatedAttribute`s that forward to `self.store`, so a method defined on a `@wrap_kvs`-decorated class runs bound to the **inner (unwrapped)** instance — `self[k]` inside it bypasses the transform pipeline. Reproduced on HEAD.

A general fix requires rethinking method delegation (e.g. rebinding delegated methods to the wrapper, guarding recursion) — a separate, more invasive change from the conditioning work, so it stays open. Documented in `misc/docs/dol_architecture_map.md` §5.5 and the `dol-dev-wrap-kvs` skill. `priority: high`, Wave 1.

**Comment by thorwhalen** [comment]
## Design + Phase-1 fix: `wrapped_self()`

Full write-up: `misc/docs/dol_issue18_design.md` (PR #74). Code: **PR #73**.

### Root cause (confirmed)
dol wraps by **delegation (has-a)**: `wrap_kvs(SomeClass)` returns a `Store` subclass that *holds* an instance of `SomeClass` in `self.store`. A method **defined on `SomeClass`** is copied on as a `DelegatedAttribute` and runs **bound to the inner, unwrapped store**, so `self[k]` inside it bypasses the transform pipeline. It **compounds** under stacked/`Pipe` wraps. (Distinct from #9; shares its root with #6.)

### Blast radius (26 ecosystem sites surveyed)
**0 break · 6 latent bugs a fix resolves · 2 neutral · 18 immune.** The 6 are *real* bugs today, e.g. `xdol.PyFilesReader(asyncio).is_pkg()` returns `False` for an actual package (membership tests the un-relativized inner keys), and a decoy `.cfg`/`.bak` leaks past `SetupCfgReader`/`PyprojectReader`'s `filt_iter`. So this is a correctness fix, not just ergonomics.

### The fix (Phase 1, backward-compatible)
```python
from dol import wrapped_self

@wrap_kvs(...)                 # or filt_iter / mk_relative_path_store / Pipe
class MyStore(...):
    def method(self, k):
        return wrapped_self(self)[k]   # outer, transform-applying store
```
`wrapped_self(obj)` returns the outermost wrapping `Store` (climbs stacked/`Pipe` layers), or `obj` unchanged on a direct `Store` subclass / plain object (safe no-op). Implemented as an `id`-keyed weakref registry populated in `Store.__init__` + `__setstate__` (survives pickle). **Zero default behavior change** — dependents test-gate stayed green (0 PASS→FAIL). The older `sq(self)[k]` re-wrap still works.

### Why not rebind `self` to the wrapper?
Considered and **rejected** (design doc §4): under `Pipe`/stacked wraps it binds to the *innermost* layer (4 of the 6 sites stay broken; stacked-codec **writes silently corrupt**) and it crashes on `super()`/dict-native ops (`descriptor '__getitem__' for 'dict' objects doesn't apply to a 'Wrap' object`) — trading one silent bug for a subtler one.

### Terminal direction
The clean structural fix is **is-a wrapping** (make `wrap_kvs(Class)` a real subclass), which would resolve **#18 and #6 together** — deferred to a major version, opt-in first, gated on a green dependents run. `wrapped_self` becomes a no-op then, so shipping it now forecloses nothing.

Leaving this open until Phase 1 (#73) merges; will note the is-a follow-up on #6.

**Comment by thorwhalen** [comment]
### Phase 1 shipped + Phase 2 (downstream fixes) up

**Phase 1** (`wrapped_self`) merged in **dol 0.3.50** (PR #73). Adversarial review during the PR caught and fixed a `copy.copy`+GC edge case (multi-valued backref registry + identity-guarded climb). Full ecosystem gate: 39/56 dependents pass, **0 regressions** (the rest fail identically at baseline).

**Phase 2** — the 6 latent-bug sites the design flagged are now fixed downstream via `wrapped_self(self)`:
- [i2mint/xdol#4](https://github.com/i2mint/xdol/pull/4) — `PyFilesReader.is_pkg`/`init_file_contents` + `SetupCfgReader`/`PyprojectReader.dependencies_from_all` (with decoy-file regression tests)
- [i2mint/unbox#2](https://github.com/i2mint/unbox/pull/2) — `ModuleNamesImportedByModule.print_kvs`
- [thorwhalen/lexis#2](https://github.com/thorwhalen/lexis/pull/2) — `WordnetElement.dict_of_non_empty_values`

**Phase 3** (terminal, deferred to a major version): the structural is-a wrapping that would resolve #18 **and #6** at the root — see `misc/docs/dol_issue18_design.md`. Keeping this open to track that.

**Comment by thorwhalen** [comment]
Status update (verification pass, not a close): **Phase 1 has shipped; Phases 2 and 3 have not.** Leaving this open.

**Shipped (PR #73, design in PR #74 / `misc/docs/dol_issue18_design.md`):** `dol.wrapped_self` — a weakref-backed inner-store -> wrapper back-reference registered in `Store.__init__` (and re-registered on `__setstate__`), plus a `wrapped_self` property on `Store`. Inside a method of a `wrap_kvs`-wrapped class, `wrapped_self(self)[k]` now recovers the outer, transform-applying view, which is exactly the asymmetry reported here. It is exported from `dol/__init__.py`, doctested in `dol/base.py`, and covered in `dol/tests/base_test.py`. Default behaviour is unchanged, so this is additive.

**Still open, per the design doc:**
- *Phase 2* — convert the known benefit sites in the dependents to `wrapped_self(self)` (the design doc names 6: one in `lexis`, one in `unbox`, four in `xdol`), each with a regression test. Those are follow-on PRs to the dependents, not to `dol`.
- *Phase 3* — the is-a wrapping mode (`_StoreTransformMixin` + subclass builder behind an opt-in `mode="isa"`), which is the structural fix that makes `self` genuinely be the wrapped store rather than needing a lookup. Only after that does `wrapped_self` become deprecated.

So the reported symptom now has a supported workaround/blessed pattern, but the underlying has-a design that causes it is still in place.

**Comment by thorwhalen** [comment]
## `wrapped_self` has a silent failure mode: it degrades to the raw leaf when the wrapper is unreferenced

Found while building an S3 adapter on the pattern this issue blessed. `misc/docs/dol_issue18_design.md`
recommends `wrapped_self(self)` as the Phase-1 answer, and `xdol`, `unbox` and `lexis` have
adopted it — so they inherit this.

### The hole

A `DelegatedAttribute` returns a method bound to the **leaf**, holding no reference to the
wrapper. So when nothing else holds it, the wrapper is freed before the method body runs, its
weakref dies, and `_register_wrapper_backref`'s cleanup callback **pops the `id(inner)` key** —
leaving a state indistinguishable from "never wrapped".

```python
from dol import KeyCodecs, wrapped_self
from dol.base import KvReader
from dol.dig import inner_most_key

class BR(KvReader):                       # a leaf that owns a prefix
    def __init__(self, prefix=''): self.prefix = prefix
    def _id_of_key(self, k): return f"{self.prefix}{k}"
    def _key_of_id(self, i): return i[len(self.prefix):]
    def __iter__(self): yield from ()
    def __getitem__(self, k): return b'v'
    def m_abs(self, k): return inner_most_key(wrapped_self(self), k)

s = KeyCodecs.prefixed('x/')(BR('logs/'))
s.m_abs('b.txt')                                   # 'logs/x/b.txt'   correct
KeyCodecs.prefixed('x/')(BR('logs/')).m_abs('b.txt')
                                                   # 'logs/b.txt'     WRONG, silently
```

Three things make this worse than it looks:

1. **The wrong answer is a plausible `str`.** When the leaf owns a prefix, its own `_id_of_key`
   still fires and produces a well-formed key addressing the wrong object. A leaf *without*
   `_id_of_key` would at least return `None`. So a type check does not catch it.
2. **The predicate is "no live strong reference", not "temporary".** A temporary caught in a
   reference cycle silently starts working, and `operator.methodcaller('m', k)(obj)` is correct
   where `obj.m(k)` is not. The failure is **intermittent**.
3. **Reproduced on CPython 3.10, 3.11, 3.12, 3.13, 3.14.** Not gc-, version- or
   bytecode-specific; it is refcount-driven, and `gc.disable()` changes nothing.

`dol`'s own tests do not cover it — every case in `dol/tests/base_test.py` binds the wrapper to
a name and therefore passes.

### It *is* detectable, at least

`Store.__init__` probes the leaf with `hasattr(self.store, "KeysView")`, so a leaf can record
that it was ever wrapped and refuse when `wrapped_self(self) is self` but that flag is set.
Verified against instance-wrap, class-wrap and `Pipe`. That is a hack riding on an incidental
probe, and it misses the unpickle path (`__setstate__` re-registers without going through
`__init__`), but a two-line change to `_register_wrapper_backref` — stamp `id(inner)` in a
plain set alongside the weakref — would do it properly. `inner` is provably alive from
registration through the call (the bound method holds `__self__`), so the id cannot have been
recycled.

Possible directions, none of them mine to choose: have `DelegatedAttribute.__get__` return a
wrapper-retaining bound method; keep a strong reference for the duration of the call; or land
is-a wrapping, which removes the registry entirely.

### Relationship to #83

#83 is the **inverse direction** of this same has-a root: there, a method *receives* an unmapped
key; here, a method's `self[k]` bypasses the transforms. Is-a wrapping resolves both — which is
the strongest argument for it being the terminal direction.

Not blocking anything of mine: [s3dol ADR-0011](https://github.com/i2mint/s3dol/blob/master/misc/docs/decisions/0011-keyed-capability-surface.md)
routes around `wrapped_self` entirely by exposing capabilities as sibling stores keyed through
`__getitem__`, which needs no key-resolution primitive. Reporting it because the documented
remedy for a *No Silent Failures* project currently has a silent failure.
